### ISSUE TITLE
Error on ipfs add --hash=sha1, but allow with --force

### ISSUE THREAD

**Issue by lidel** [ipfs/go-ipfs#8703]
### Checklist

- [X] My issue is specific & actionable.
- [X] I am not suggesting a protocol enhancement.
- [X] I have searched on the [issue tracker](https://github.com/ipfs/go-ipfs/issues?q=is%3Aissue) for my issue.

## Problem

> Suggestion extracted from https://github.com/ipfs/go-ipfs/issues/8650

The default is SHA2-256, but we support SHA1 for legacy reasons.
SHA1 was broken, and systems like git moved away from it, but we are not warning user when they opt-in to using it.

## Solution

We can't remove support for it, because  there should be a way for people to migrate the data.

There should be a list of "legacy/deprecated" hash functions like SHA1 which would be discouraged:

```console
$ ipfs add --hash=sha1 file
Error: selected hash function is no longer secure; use --hash=sha2-256 or pass --force

$ $ ipfs add --hash=sha1 file
Added ...
```

**Comment by aschmahmann** [comment]
Basically all the insecure hash functions are already forbidden by https://github.com/ipfs/go-verifcid/blob/46876f413195d36227ee5c7d6f16e995b62f754d/validate.go#L15. The only exception here is SHA1 because at the time it was deemed valuable (and I assume insufficiently broken) https://github.com/ipfs/go-verifcid/blob/46876f413195d36227ee5c7d6f16e995b62f754d/validate.go#L30.

This means that at the moment this really only applies to SHA1.

-----

We could also just make using SHA-1 safer by utilizing the cryptoanalytic "fix" used in places like Git https://github.com/cr-marcstevens/sha1collisiondetection.

To do this there are two pieces:
1. Make https://github.com/ipfs/distributions be able to build with CGO
2. Have go-ipfs use something like https://github.com/aschmahmann/go-sha1collisiondetection (CGO wrapper around the C library with the cryptoanalytic fix). Work on the wrapper started in a PR to that repo.

Related: https://github.com/multiformats/go-multihash/issues/57

**Comment by Jorropo** [comment]
3. Rewrite or machine translate the cryptoanalytic fix in Go.

I guess it's mainly just doing arithmetic on byte buffers, this should be trivial to machine translate (and lift up manually).

**Comment by aschmahmann** [comment]
> We should avoid relying on CGO for core features.

I agree it'd be nice to, but I also suspect this is something we're going to run into wanting anyway. There is so much code written with FFI that restricting ourselves from using any of it seems rough, as opposed to evaluating the code and seeing if the CGO costs are worthwhile.

> Rewrite or machine translate the cryptoanalytic fix in pure Go.

Not sure how painful this is. When I was looking around at this I didn't find native implementations in any other language which makes me think maybe it's painful to do given how important it is to have the implementation correct.

If we can get someone to write and review the SHA-1 code in Go that's fine, but IMO blocking on this seems like a bad idea that's prevented this from being integrated earlier (i.e. the linked go-multihash issue is from 2017)

**Comment by hsanjuan** [comment]
This is marked as P1, so probably something should be done. I think we can just remove the SHA1 exception in verify-cid and forget. I don't see why we would spend time integrating collisioncheck tbh. Probably no one is using sha1 in ipfs land anyways, and wouldn't feel sorry for those that do.

**Comment by aschmahmann** [comment]
Example of someone trying to use this recently https://github.com/NixOS/nix/issues/859#issuecomment-3044910842. Generally speaking use cases with SHA-1 and IPFS that I see still tend to revolve around Git, which has still not yet moved to SHA-2 by default yet (and major platforms like GitHub don't even support it as an option).

While it'd be nice and simplifying to only support hash functions that are deemed cryptographically safe idk if the removal of SHA-1 support is really the move right now, or if doing the deprecation / removal here will just result in wasted time related to breaking existing users as happened with https://github.com/ipfs/kubo/issues/9717.

**Comment by guillaumemichel** [comment]
Triage notes:
* Blocked until git 3 is released
